### PR TITLE
API: Recognize MSSQL data source URLs

### LINKED ISSUE #25067: Cannot connect to SQL Named Instance
**What happened**:
When trying to connect to a named instance when using version 7.0 we cannot specify a named instance. When using a backslash (`\`) to seperate the server name from the instance name, it gives us an error. See attached.

**What you expected to happen**: We should we able to connect as in the past to the server. This worked in version 6.7.2

**How to reproduce it (as minimally and precisely as possible)**:
Try to create a connection to a new instance using server\InstanceName. Is seems that you interpret each single slash (back slash) as two slashes.

**Anything else we need to know?**: No

**Environment**:
- Grafana version: 7.0
- Data source type & version: SQL Server
- OS Grafana is installed on: CentOS
- User OS & Browser: Chrome

- Grafana plugins:
- Others:

ServerName
Connection

### PR DESCRIPTION
**Pull request by aknuds1**
**What this PR does / why we need it**:
Implement MSSQL specific data source URL validation logic.

**Which issue(s) this PR fixes**:
Fixes #25067.

**Special notes for your reviewer**:
I read the MSSQL [driver](https://github.com/denisenkom/go-mssqldb) documentation in order to figure out valid connection strings. I figure I should probably add support for the two other MSSQL URL formats too, but starting out with this.

### CODE REVIEW

**Comment by aknuds1** [comment]
@reuvygroovy Can you please test this PR and see if it works for you?

**Comment by reuvygroovy** [comment]
How can I test it? What should I install/replace?

**Comment by aknuds1** [comment]
@reuvygroovy Are you able to build Grafana from source?

**Comment by reuvygroovy** [comment]
No, haven't done that yet. If you explain how though I am willing to try.

**Comment by aknuds1** [comment]
Read the [developer guide](https://github.com/grafana/grafana/blob/master/contribute/developer-guide.md).

**Review by papagian — APPROVED** [review]
LGTM
It would be nice to have some tests for the different cases that `parseURL` supports.
Also, off the scope of this one, ideally each datasource should have each one validation method (by implementing the same interface) but this requires a lot of refactoring.

**Comment by aknuds1** [comment]
@papagian Those tests are implemented via TestNewDataSourceProxy_MSSQL. If mssql.ParseURL has a bug it should be caught there. It's preferable to test as few implementation details as possible.

**Comment by papagian** [comment]
> @papagian Those tests are implemented in TestNewDataSourceProxy_MSSQL. If mssql.ParseURL has a bug it should be caught there. It's preferable to test as few implementation details as possible, to reduce maintenance/cognitive load.

correct! forget about my comment!

File: pkg/api/pluginproxy/ds_proxy_test.go [hunk 1]
```diff
@@ -583,11 +584,62 @@ func TestNewDataSourceProxy_ProtocolLessURL(t *testing.T) {
...
 	cfg := setting.Cfg{}
 	plugin := plugins.DataSourcePlugin{}
+
 	_, err := NewDataSourceProxy(&ds, &plugin, &ctx, "api/method", &cfg)
 
 	require.NoError(t, err)
 }
 
+// Test wth MSSQL type data sources.
+func TestNewDataSourceProxy_MSSQL(t *testing.T) {
```
Comment (marefr) [inline pkg/api/pluginproxy/ds_proxy_test.go:594]: Testing the proxy with MSSQL doesn't make sense.

Comment (aknuds1) [reply pkg/api/pluginproxy/ds_proxy_test.go:594]: @marefr Why is that a problem? The proxy supports data sources of MSSQL type, same as the handlers for adding and updating data sources. The benefit of testing from this level is that you're testing fewer implementation details, and you can see that the chain of code is working.

Comment (aknuds1) [reply pkg/api/pluginproxy/ds_proxy_test.go:594]: @marefr OK after our conversation I understand what you mean. I'll try to look into it.

**Comment by aknuds1** [comment]
@reuvygroovy Can you try the latest [nightly](https://grafana.com/grafana/download/7.1.0-96b2dad4pre)? It should have the fix.

**Comment by reuvygroovy** [comment]
Works! Awesome! Thanks :-)

**Comment by aknuds1** [comment]
@reuvygroovy Great!

**Comment by gitszakal** [comment]
Hello Team,
Unfortunately I have the same issue.

I have an SQLExpess database and a Windows installed Grafana 7.5.4. on the same PC.
Host: localhost\SQLEXPRESS:1433 brings this failure:
_Validation error, invalid URL: "localhost\\SQLEXPRESS:1433"_
With double \\ the host I get this:
_Validation error, invalid URL: "localhost\\\\SQLEXPRESS:1433"_
I have tried to add data source PostgreSQL, but I get the same issue.
Just with name SQLEXPRESS:1433 brings this failure:
_dial tcp: lookup SQLEXPRESS: no such host_

I am not an experienced programmer, could you please help me on this, explaining on an easy level? :)
Thank you in advance

**Comment by reuvygroovy** [comment]
Did you try computer name instead of localhost?

Also try

.\SQLExpress

Try also without port

**Comment by gitszakal** [comment]
Thank you for your willing to help :)
Yes, I tried with computer name, with, without port, but always the same.
.\ I haven't tried before, but now this brings the same message as well:
Validation error, invalid URL: ".\\\\SQLEXPRESS:1433"

**Comment by reuvygroovy** [comment]
can you connect with Management Studio on the same PC? on a different computer?

**Comment by gitszakal** [comment]
Connection to the database on the same PC works with Management Studio.
I haven't tried from different computer, but maybe I can try to find a way to try... Does it makes any difference on the host name connection?

**Pull request merged**